### PR TITLE
Mef reader

### PR DESCRIPTION
**Pull request by jiecui**
Hi Robert,

We changed the code as you advised.  Please test code with `test_mayo_mef.m`.  To do this, you will need to go to the [GitHub of MultiModal NeuroImaging Lab](https://github.com/MultimodalNeuroimagingLab/mef_reader_fieldtrip) to download **mef_reader_fieldtrip** package.  Unzip it if necessary and look for the folder 'mayo_mef' under 'fieldtrip/external'.  Then put it under the 'external' directory of your FieldTrip.

Please let us know if there is any issue of the code.

Thank you,
Richard
--
MNL Group
Mayo Clinic Rochester
MN, USA

### CODE REVIEW

**Comment by jiecui** [comment]
Hi Robber,

Just wonder whether you got a chance to test the code?  I also have a question for you.  In your last reply (pull request #1375), you said there should not be any binary data in the external toolbox.  I understand that the binary data for testing the code should not be included.  How about the binary code from C-MEX functions, which are essential to the toolbox?  Can we include the binary data of C-MEX functions?

Please just let us know if you have questions regarding the code.

Many thanks,
Richard
----
MNL Group
Mayo Clinic Rochester
MN, USA

**Comment by robertoostenveld** [comment]
Hi @jiecui 
I have been very busy the last few weeks with the FieldTrip toolkit workshop, which had to be transformed from the regular format to an online course. The online toolkit workshop happened last week; it was a great success. I still have to wrap some things up and will then return to this.

**Comment by jiecui** [comment]
Hi @robertoostenveld, Thanks for the message!  Yes, I'm aware of the workshop, which is online now.  Congratulations!

**Comment by schoffelen** [comment]
Dear Richard, allow me to chime in a bit here. Thanks for your efforts so far, and great to see that there's interest in getting your fileformat implemented in FieldTrip. I think that's a great service to the community. 

Adding c-source code is in principle OK (because it's not a lot of data). However, we typically try to avoid mex-files as much as possible. I saw on github that you only have compiled versions for MacOS and windows, and users would probably expect that there's a compiled linux version as well. If you cannot provide this, it should be properly documented, and also some pointers should be given as to how people can compile it themselves. We have a few mex-files in the core-fieldtrip package, which we try and maintain as good as possible, but we try to stay away from taking responsibility for the externally contributed files. 

Do the mex-files in your package have a slower (fallback) corresponding functional m-file?

**Comment by MaxvandenBoom** [comment]
Hi all!

@schoffelen No, unfortunately there is no pure matlab code currently available to do this. The package uses a compact underlying c library dedicated to the MEF file-format.

I understand that you would typically try to avoid mex-files. What is the typical way to go about this in Fieldtrip. When a users tries to load a file-format, should the file format extension then compile or download on-demand?

**Comment by jiecui** [comment]
Hi Schoffelen and Max,

Thank you both for the comments.  For now, the package of our pull request doesn't include any c-mex source code and/or pre-compiled binary code.  We followed Robert's instructions to include only the basic functions of ft_read_header, ft_read_data, ft_toolbox and a test script in this pull request.  To test the code, one needs to download the external package from [GitHub of MultiModal NeuroImaging Lab](https://github.com/MultimodalNeuroimagingLab/mef_reader_fieldtrip).  We hope Robert or a member of Fieldtrip team can test the toolbox, and after that we will deploy it according to Fieldtrip's protocol.

Thank you,
Richard
----
MNL Group
Mayo Clinic Rochester
MN, USA

File: fileio/ft_filetype.m [hunk 1]
```diff
@@ -1425,10 +1459,6 @@
   type = 'motion_c3d';
   manufacturer = 'https://www.c3d.org';
   content = 'motion capture data';
-elseif filetype_check_extension(filename, '.mvnx') && filetype_check_header(filename, '<?xml')
```
Comment (robertoostenveld) [inline fileio/ft_filetype.m:1428]: It is not clear to me why this piece of code has been deleted

File: fileio/ft_filetype.m [hunk 2]
```diff
@@ -1456,10 +1486,6 @@
   type = 'maus_textgrid';
   manufacturer = 'MAUS/WebMAUS';
   content = 'segmented text';
-elseif filetype_check_extension(filename, '.fcsv')
```
Comment (robertoostenveld) [inline fileio/ft_filetype.m:1459]: It is not clear to me why this piece of code has been deleted

File: fileio/ft_read_data.m [hunk 3]
```diff
@@ -1026,32 +1039,11 @@
     % recordings. The code currently concatenates these trials.
     % We could set this up as separate "trials" later.
     % We could probably add "boundary events" in EEGLAB later
-
```
Comment (robertoostenveld) [inline fileio/ft_read_data.m:1029]: It is not clear to me why this recent update to the code has been reverted

File: fileio/ft_read_event.m [hunk 4]
```diff
@@ -19,6 +19,8 @@
 %   'tolerance'      tolerance in samples when merging Neuromag analogue trigger channels (default = 1, meaning that an shift of one sample in both directions is compensated for)
 %   'blocking'       wait for the selected number of events (default = 'no')
 %   'timeout'        amount of time in seconds to wait when blocking (default = 5)
+%   'password'       password structure for encrypted data set (such as mayo_mef30 and mayo_mef21)
+%   'sortchannel'    = sort channel order according to either alphabet or number (default is alphabet if header is not provided, otherwise use header.SortChannel)
```
Comment (robertoostenveld) [inline fileio/ft_read_event.m:23]: channels are not part the output of ft_read_event, so I do not understand why you would want to sort on them

File: fileio/ft_read_header.m [hunk 5]
```diff
@@ -15,6 +15,8 @@
 %   'coordsys'       = string, 'head' or 'dewar' (default = 'head')
 %   'chantype'       = string or cell of strings, channel types to be read (NeuroOmega, BlackRock).
 %   'headerformat'   = name of a MATLAB function that takes the filename as input (default is automatic)
+%   'password'       = password structure for encrypted data set (such as mayo_mef30 and mayo_mef21)
+%   'sortchannel'    = sort channel order according to either alphabet or number (default is alphabet)
```
Comment (robertoostenveld) [inline fileio/ft_read_header.m:19]: the `sortchannel` option allows the output of ft_read_header and ft_read_data to become inconsistent if the user were to use it with one, and not the other. Sorting channels is trivial to do outside of ft_read_header and ft_read_data and therefore I don't see the reason to implement it as part of the functions. If anywhere, it would make more sense to implement it in ft_preprocessing.

File: fileio/private/ft_hastoolbox.m [hunk 6]
```diff
@@ -408,9 +410,7 @@
     dependency = {'bf.ttest', 'bf.ttest2'};
   case 'EZC3D'
     dependency = {'ezc3dRead', 'ezc3dWrite'};
-  case 'XSENS'
```
Comment (robertoostenveld) [inline fileio/private/ft_hastoolbox.m:411]: why did you remove the XSENS support?

File: fileio/ft_read_data.m [hunk 7]
```diff
@@ -138,6 +140,8 @@
 dataformat      = ft_getopt(varargin, 'dataformat');
 chanunit        = ft_getopt(varargin, 'chanunit');
 timestamp       = ft_getopt(varargin, 'timestamp');
+password        = ft_getopt(varargin, 'password', struct([]));
+sortchannel     = ft_getopt(varargin, 'sortchannel', '');
```
Comment (robertoostenveld) [inline fileio/ft_read_data.m:144]: looking at the code further down, I see that sortchannel only works for MEF and not for any other file format. That is inconsistent with the documentation and will confuse users

**Comment by robertoostenveld** [comment]
There is a serious issue with this pull request, since it inadvertently reverts some recent improvements to FT code that are not related to the MEF format. 

What I suggest is that I will test it like this nevertheless, but @schoffelen please note that this PR cannot be merged, even if the MEF specific test is OK.

File: fileio/ft_filetype.m (hunk 2 above)
Comment (jiecui) [reply fileio/ft_filetype.m:1459]: Hi @robertoostenveld, No, we didn't intend to change other functions, except ft_read_header, ft_read_event and ft_read_data in this PR.  I guess this issue is because we sent the PR on Apr. 17th, and there are modification of the master branch of Fieldtrip since then.

File: fileio/ft_filetype.m (hunk 1 above)
Comment (jiecui) [reply fileio/ft_filetype.m:1428]: Hi @robertoostenveld, Please see the comment below. Thanks!

File: fileio/ft_read_data.m (hunk 3 above)
Comment (jiecui) [reply fileio/ft_read_data.m:1029]: Please see the comment above

File: fileio/ft_read_event.m (hunk 4 above)
Comment (jiecui) [reply fileio/ft_read_event.m:23]: We'll discuss whether to keep 'sortchannel' option in ft_read_event. Thanks.

File: fileio/ft_read_header.m (hunk 5 above)
Comment (jiecui) [reply fileio/ft_read_header.m:19]: That's a good point.  A mechanism to make consistent between `ft_read_header` and `ft_read_data` has already built in the code.  Sorting channel according to channel number is of practical importance and is not that straightforward, as the user needs to go deep into the fields of header.  However, we'll discuss whether to sort the channels in `ft_read_header` and `ft_read_data`.

File: fileio/private/ft_hastoolbox.m (hunk 6 above)
Comment (jiecui) [reply fileio/private/ft_hastoolbox.m:411]: No, we didn't try to do that.  Please see the comments above.

File: fileio/ft_read_data.m (hunk 7 above)
Comment (jiecui) [reply fileio/ft_read_data.m:144]: Yes, we'll discuss the option of `sortchannel`.

**Comment by jiecui** [comment]
Hi Robert @robertoostenveld, 

Thanks for the comments and testing.  We'll discuss and perhaps open a new pull request.

**Comment by robertoostenveld** [comment]
I have re-implemented the suggested code changes in a clean branch. See https://github.com/robertoostenveld/fieldtrip/tree/mef_reader 

It does not run through completely, but with the test scripts that you provided I can load the data. Can you please check the test_mayo_mef script and suggest code changes to make thaht to work on my development branch (i.e. not on master from fieldtrip/fieldtrip)?

**Comment by jiecui** [comment]
OK, will do.  Thanks!
